### ISSUE TITLE
Github Desktop vs Git + Terminal

### ISSUE THREAD

**Issue by ChrisOwen101** [CodeYourFuture/syllabus#77]
**Github Desktop vs Git + Terminal**

Before we start working on the ticket about designing a Git lesson we need to have a discussion on which method is best for us to teach.

We maybe also decide to design the lessons in such a way that we can leave it up to the school to define.

**Overview**

Git is one of the key tools we use to submit homework and historically has been tricky getting students setup with. Last year in London we chose to teach [Github Desktop](https://desktop.github.com/) and Git Theory rather than trying to teaching Terminal which could be viewed as a barrier to entry that increase cognitive load.

This was met with moderate success - students were able to submit homework but the majority still had problems submitting homework and doing basic Git Tasks. Some issues were reduced including

- Typing and Grammatical Errors in commands
- Memory recall issues with forgetting commands

However it did introduce issues

- The nuts & bolts of git are obscured a lot meaning fundamentals understands of the tool are not formed so quickly
- Students did not have good Terminal Skills by the time they reached React which caused problems.

**Success Metrics**

This lesson will be a success if student can 

- define what Git is and what it does
- define what the differences between Git and Github are
- create a repository on Github
- clone a repository using Github Desktop or Command Line
- make changes to a repository using Github Desktop or Command Line
- create and move between branches using Github Desktop or Command Line

Students will be using Github extensively to submit homework.

**Discussion**

Given the above - we should decide which we want to teach - `Github Desktop` vs `Git + Terminal`

Below, could we list out the Positive & Negative points for each solution so we can be an accurate decision.

Do not be afraid to be opinionated one the thread - let's work out the best solution.

**Comment by IsmaelMartinez** [comment]
Probably worth getting some of the info that is already in the https://github.com/CodeYourFuture/html-css-git-exercises/tree/master/week-1/9-git-branch course. I personally think that you should teach the terminal option 1st.

**Comment by ChrisOwen101** [comment]
**From Javier Lopez**

If the goal is to teach git. It is better to understand git first and then it is easier to teach any front end (Github Desktop/VS plugin/IJ or anything else). Usually if you teach a front end first, people (students and professions alike) rely on the front end tool without understanding the tool. If the goal is to have them submit the homework correctly, without knowing the details of git, a front end makes sense.
Learning to use a terminal to me is a separate issue from learning git. This is a required skill they will be forced to learn eventually. In my experience, most people (even professionals) learn the bare bones just to get by. But this is a very complicated topic on its own.

**Comment by ChrisOwen101** [comment]
**From Mitch Lloyd**

I completely agree with @Javier on this point. I think understanding of the git CLI is essential.
I have lots of experience teaching Git to beginners and think that it is possible to have a blend of high level conceptual understanding of Git together with some demos on the most important commands.
I think the resources you shared with me the other day @Chris Owen are excellent and would really help to facilitate this.

**Comment by ChrisOwen101** [comment]
**From Mike Hayden**

I think git is pretty complex to get your head around at first and anything visual that can be used to help that understanding will benefit the student, so I would probably lean more toward Git Desktop or VS Code's git tools.
However I think adding in basic git terminal commands to supplement the lesson would be good as a way to introduce the terminal or to augment what terminal commands the students have already been exposed to at this point in the course.
The more they can see the terminal as a way to send commands to programs the more they will understand its power and become more comfortable with it. This will be especially useful when they start to use Webpack and other CLI tools.
I think @Chris Owen said that there is a whole day set aside for git now, which might be able to accommodate mostly Git Desktop but with the introduction of some git terminal commands.

**Comment by ChrisOwen101** [comment]
**From Daniel Wagner-Hall**

Some other concerns not listed on the issue:
 1. In London we found that teachers/mentors were less able to help in GitHub Desktop than on the command line because of their own unfamiliarity. If we choose command line, this goes away. If we choose Desktop, we should make sure our material (perhaps a cheat-sheet) is useful to volunteers too
2. A corollary of teaching the command line is teaching enough of less that people know how to navigate and exit things like git log, and teaching enough of vim that people know how to stop editing a commit message when they forgot to pass -m to git commit and don't have their $EDITOR set up well. These are not great things to need to teach...

**Comment by ChrisOwen101** [comment]
**From Javier Lopez**

That's a great point. Mentors tend to use different front ends.

**Comment by ChrisOwen101** [comment]
**From Daniel Wagner-Hall**

That said, I think the command line is the way to go. My thoughts are:
1. Sometimes when students get stuck, we need them to drop to the command line to fix their situation, and it would be nice if this was building in learning, rather than not-understood magic
2. It's often easier to StackOverflow a command line answer than a GitHub Desktop answer, and sometimes things busy aren't possible from GD
3. It's a more general skill, and will help them with other command line things in the future
I completely agree that we should teach the idea before the syntax.
Going past this discussion a bit (maybe too off topic for now):
When we get to teaching the syntax, I think ideally we would have them write a story with the person next to them, in a plain text file, to separate the concept from code. Then, have them publish a really simple static HTML page to GitHub pages or netlify or similar. And finally, have them submit some "homework" in-class, have our teachers update the homework repo, and have the students pull that in.

**Comment by ChrisOwen101** [comment]
**From Loic Le Henaff**

Some feedback from CYF Glasgow here :slightly_smiling_face:. We never taught GitHub Desktop, we always taught how to use the terminal with the basic commands and how to fully rely on Git command line for doing everything. And this worked pretty well so far, because we tried to teach the core concepts of Git first before even considering the tools like GitHub Desktop which may hinder the teaching. And to be honest, personally i’ve never had to use GitHub Desktop, and i don’t know many devs who systematically use a GUI over the CLI, often because the CLI is just faster and more convenient over a GUI. So even in terms of getting the students more prepared for a professional experience, using Git in the command line is better. If they understand Git with the command line, they can easily learn whatever desktop tools they want if they need.

**Comment by ChrisOwen101** [comment]
**From Loic Le Henaff**

In Glasgow, we didn’t follow the intro to coding course in the master branch this year. Here are the materials we used in Glasgow:
CYF Scotland intro course: https://codeyourfuture.github.io/syllabus-scotland/introduction/
Git course: https://codeyourfuture.github.io/syllabus-scotland/introduction/week-2/lesson.html#git
Here is the workshop transcript @R used to introduce Git: https://codeyourfuture.github.io/syllabus-scotland/introduction/week-2/git-workshop-transcript.html
And the video @R created for students to see at home after the class: https://www.youtube.com/watch?v=bxaF0tJ1u8c

**Comment by ChrisOwen101** [comment]
**From Atanas Kozarev**

I lean towards what @Mike Hayden is saying that GitHub desktop is user-friendly. And while it’s essential to know the command line, in which priority do you rank all the other essential things. Can you get by using the desktop client for a while and still are able to function in a team or to manage your CYF homeworks and own repos?
We also have 37 students in Birmingham. It’s a challenge of different magnitude. You have the same time but you have to double-tripple output vs e.g. Glasgow, as far as I know. When there are problems that’s double the troubleshooting.
To use the terminal, you first have to understand folder structure and basic navigation commands. To many Windows Desktop is your starting place, it’s well buried under C:/Users/Myname/Desktop etc etc I have watched some of our students try and use the terminal for the first few weeks and it’s hard. How does the OS GUI experience relate to the terminal paths and commands. List, change folder, delete potentially. We have to teach those things at the same time?
The GitHub desktop has convenient buttons - like open with VS Code, view in Explorer, view on Github and other useful click buttons.
I don’t see how using the Desktop client can disadvantage anyone immediately.
I agree that one gets accustomed to it and perhaps reluctant to then learn terminal.

**Comment by ChrisOwen101** [comment]
**From Loic Le Henaff**

Indeed it’s a lot of students that you have in Birmingham, which raises different issues. In Glasgow we’ve limited on purpose the number of students we’re taking every year to max 15 (even though we received more than 100 applications last year), simply because given the number of mentors we didn’t feel we could support more students effectively. To come back to Git, I do think (but again personal opinion) that learning GitHub Desktop rather than Git CLI disadvantages students. The learning is in Git, not in the GUI tool. Same argument for using the OS, the learning is in the terminal, not in the Windows explorer. It’s definitely not easy to teach these concepts, but i think it’s the right way for people who wish to become professional software developer. There are good chances they’re not going to work with Windows and that their future company won’t be using GitHub Desktop either, that’s why teaching the underlying technology rather than the tools is far more important to me. This knowledge can then be adapted and reused because it’s the core concepts which are taught, which is not dependent on a given software and OS platform.

**Comment by ChrisOwen101** [comment]
**From Chris Owen**

For my input and continuing vaguely from Loic & Atanas points -
At the highest level points in defence of Github Desktop are around cognitive load and the real importance of teaching Git on the course. We use Git at CYF because (in order of importance)
We've not found a better tools to do Code Reviews with yet
Github is an industry standard and students should have a portfolio of work on it
Git is an industry standard and students should be able to use it
We all use Git and Github
I'd argue that we can achieve 1 & 2 without touching Git itself and that we should use this to our benefit - our students are already overloaded with the amount that we give them in the first 6-12 weeks of the course and I don't think Git is the best use of their time during this stage. If given the choice on a students using precious time to work out how to use Git and spending more time practising JavaScript, I'd prefer JavaScript.
What I'm not saying is that Git & Terminal is an irrelevant skill - quite to the contrary - it's just not the right time to be teaching a tool that has functionality far beyond what they will actually be using it for. We should endeavour to provide students with the tools to solve problems they are currently understanding.
The learning is in Git, not in the GUI tool.
To take this point alone - I absolutely agree - however what we're trying to achieve is not to teach Git. It's to provide the students with the tools they need to solve the problem in front of them, in this case, to submit their homework.
(I'm not absolutely convinced one way or the other yet - but I'm enjoying this debate and lots of the Pro-Terminal approach have been represented already.)
(p.s. Thanks for sharing your resources @Loic - super helpful!)

**Comment by ChrisOwen101** [comment]
**From Javier Lopez**

> however what we’re trying to achieve is not to teach Git. It’s to provide the students with the tools they need to solve the problem in front of them

If that’s the goal, teaching git does not make sense. I would use GitHub Desktop or the VS plugin. Many of my students started using that on their own as well.

**Comment by ChrisOwen101** [comment]
**From Loic Le Henaff**

If the first goal is to do code reviews, there is something much simpler than Git. We can simply ask the students to zip their files, send it to their edu mentor by email, the edu mentor adds comments on the files and send it back. It’s a workflow much nearer to what a normal user does. It’s something I did also as part of another professional training program for new hires and it worked very well. So to me, the primary reason for teaching Git is not for a tool for code reviews, simpler alternatives exist.
I don’t think the decision should be along the lines “we don’t teach Git because we can then spend more time in JS”. Git and the Terminal are skills that students can develop along side any other modules of the program, they are cross modules skills. That’s why i think it’s important to introduce them as early as possible, so students have opportunities to use and master them during the rest of the class, little by little. Again, it’s something which worked well in Glasgow in the past years. Everyone knew how to open a terminal, go to specific folder, create a project or a file, commit and push files with git, start an app etc. There is a misconception that because it’s the terminal, it’s more difficult to use than a GUI. I think we should improve the way we teach the terminal and git, find better ways to explain, find ways to relate git to things that students know, it’s what we tried to achieve here. But going all-in in graphical tools, spending time to explain how a VS plugin work etc would be a real regression in my opinion.

**Comment by ChrisOwen101** [comment]
**From Javier Lopez**

@Loic I don’t know about others, but I would not do code reviews like that. GitHub and/or GitLab tools make code reviews so much simpler for the reviewer.
With regards to teaching Git and the terminal. It is an invaluable resource, but I don’t think anyone will not get a job if there are not competent on either. I can’t remember an interview where anyone bother to ask any question about a terminal or git.  Maybe others have had a different experience on the matter.

**Comment by ChrisOwen101** [comment]
**From Javier Lopez**

@Javier Sorry, i didn’t mean we should do like this for code reviews. My point was, in my opinion, the first reason we teach git is not because it’s convenient for mentors to do code reviews. The real benefit of teaching git is for students first because git is an industry standard.
Also, i don’t think all of this is about teaching something because it’s needed for a job interview, you may not have questions about React or Express either... The goal for me is to prepare them as well as it can be to be competent software developer, and for this objective, the concepts behind git and the terminal needs to be taught correctly.

**Comment by ChrisOwen101** [comment]
**From Dom Vinyard**

I personally advocate for just GitHub desktop.

>The nuts & bolts of git are obscured a lot meaning fundamentals understands of the tool are not formed so quickly

I would contest this. I concede that the nuts and bolts are obscured a lot but i don't think it follows that the fundamentals are not formed as quickly. I'd almost say the opposite is true, i think a graphical representation can help cement some of those fundamentals.

On Daniels point

> 1. Sometimes when students get stuck, we need them to drop to the command line to fix their situation, and it would be nice if this was building in learning, rather than not-understood magic

what are these 'get stuck' things, because this feels like it might be the crux of the debate (?) - reset, anything else? Obviously not looking for a full git feature list here, but wondering what a typical student might encounter in a typical year that is not facilitated by the desktop app (edited)

**Comment by ChrisOwen101** [comment]
**From Daniel Wagner-Hall**

>what are these ‘get stuck’ things

Yeah, I think reset is the big one. Basically, the main use-case I’ve seen is:
“I accidentally committed to master and pushed it, I need to fix it”
The other use-case I’ve seen that has been fiddly, but may just be from my ignorance (cheat-sheet please!) is: “I need to pull in upstream changes because exercises got fixed/added since I cloned”

**Comment by ChrisOwen101** [comment]
**From Javier Lopez**

In many cases something happened that they need help with, and the easiest way for me to help them is to have them run git status from the command line to figure out what the state of things are.  And from there all the commands I have them run look like magic to them. Even when we had taught the basics already.
